tests/test_grid.py

```diff
@@ -531,27 +725,127 @@ def test(grid, basis, err_msg, true_avg=1):
             if isinstance(grid, ConcentricGrid):
                 # values closest to axis are never accurate enough
                 numerical_avg = numerical_avg[1:]
-            np.testing.assert_allclose(numerical_avg, true_avg, err_msg=err_msg)
+            np.testing.assert_allclose(
+                numerical_avg,
+                true_avg,
+                err_msg=str(type(grid)) + " " + str(grid.sym),
+            )
 
         M = 10
         M_grid = 23
         test(
             QuadratureGrid(L=M_grid, M=M_grid, N=0),
             FourierZernikeBasis(L=M, M=M, N=0),
-            "QuadratureGrid",
         )
         test(
             LinearGrid(L=M_grid, M=M_grid, N=0, sym=True),
             FourierZernikeBasis(L=M, M=M, N=0, sym="cos"),
-            "LinearGrid with symmetry",
         )
         test(
             ConcentricGrid(L=M_grid, M=M_grid, N=0),
             FourierZernikeBasis(L=M, M=M, N=0),
-            "ConcentricGrid without symmetry",
         )
         test(
             ConcentricGrid(L=M_grid, M=M_grid, N=0, sym=True),
             FourierZernikeBasis(L=M, M=M, N=0, sym="cos"),
-            "ConcentricGrid with symmetry",
         )
+
+    @pytest.mark.unit
+    def test_symmetry_volume_integral(self):
+        """Test volume integral of a symmetric function."""
+        # Currently, midpoint rule is false for LinearGrid made with L=number.
+        def test(grid, midpoint_rule=False):
+            r = grid.nodes[:, 0]
+            t = grid.nodes[:, 1]
+            z = grid.nodes[:, 2] * grid.NFP
+            true_surface_avg = 5
+            function_of_rho = 1 / (r + 0.35)
+            f = (
+                true_surface_avg
+                + np.cos(t)
+                - 0.5 * np.cos(z)
+                + 3 * np.cos(t) * np.cos(z) ** 2
+                - 2 * np.sin(z) * np.sin(t)
+            ) * function_of_rho
+
+            if midpoint_rule:
+                r_unique = r[grid.unique_rho_idx]
+                dr = np.empty_like(r_unique)
+                dr[0] = (r_unique[0] + r_unique[1]) / 2
+                dr[1:-1] = (r_unique[2:] - r_unique[:-2]) / 2
+                dr[-1] = 1 - (r_unique[-2] + r_unique[-1]) / 2
+            else:
+                dr = np.ones(grid.num_rho) / grid.num_rho
+            expected_integral = np.sum(dr * compress(grid, function_of_rho))
+            true_integral = np.log(1.35 / 0.35)
+            midpoint_rule_error_bound = max(dr) ** 2 / 24 * (2 / 0.35**3)
+            right_riemann_error_bound = dr[0] * (1 / 0.35 - 1 / 1.35)
```

Comment:
This is what I was mentioning earlier.
A line integral over rho is accurate to what we expect (within error bound of midpoint rule). Of course it would be better if the error was guaranteed to be smaller (like how Simpson's rule does better than midpoint rule).
The would be relevant to make integrals over three dimensions (rho, theta, and zeta), more accurate since those include an integral over rho.

But I'm pretty sure it's not an issue since the integrals over all coordinates that we are interested in, are volume integrals that include the Jacobian factor - which our grids are designed to integrate well.

Reply:
Yea, so if someone wants to do an integral over rho (like say integrate the surface area of each surface over rho to get volume), they should also know to multiply by the correct metric quantity so that the end result is volume